keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1235 @@
+from the startup of other applications. Let's call "main" container the app that
+will log and "logging" container the sidecar that will facilitate it.
+
+If the logging container starts after the main app, some logs can be lost.
+Furthermore, if the logging container is not yet started and the main app
+crashes on startup, those logs can be lost (depends if logs go to a shared
+volume or over the network on localhost, etc.). While you can modify your
+application to handle this scenario during startup (as it is probably the change
+you need to do to handle sidecar crashes), for shutdown this approach won't
+work.
+
+On shutdown the ordering behavior is arguably more important: if the logging
+container is stopped first, logs for other containers are lost. No matter if
+those containers queue them and retry to send them to the logging container, or
+if they are persisted to a shared volume. The logging container is already
+killed and will not be restarted, as the pod is shutting down. In these cases,
+logs are lost.
+
+The same things regarding startup and shutdown apply for a metrics container.
+
+### Problems: service mesh
+
+Service mesh presents a similar problem: you want the service mesh container to
+be running and ready before other containers start, so that any inbound/outbound
+connections that a container can initiate goes through the service mesh.
+
+A similar problem happens for shutdown: if the service mesh container is
+terminated prior to the other containers, outgoing traffic from other apps will
+be blackholed or not use the service mesh.
+
+However, as none of these are possible to guarantee, most service meshes (like
+Linkerd and Istio), need to do several hacks to have the basic functionality.
+
+Some service meshes depend on secret (like a certificate) downloaded by other
+Init container to establish secure communication between services. This makes
+the problem of ordering of sidecar and Init containers harder.
+
+### Problems: configuration / secrets
+
+Some pods use init containers to pull down configuration/secrets and update them
+before the main container gets it. Then use sidecars to continue to watch for
+changes and perform the updates and push to the main container. This requires
+two separate code paths today. Perhaps the same sidecar could be used for both
+cases.
+
+### Goals
+
+This proposal aims to:
+- make containers implementing the sidecar pattern first class citizens inside a
+  Pod
+- solve a Job completion issue when sidecars should run continuously
+- allow mixing initContainers and sidecars for a choreographed startup sequence
+- allow easy injection of sidecar containers in any Pod
+
+### Non-Goals
+
+This proposal doesn't aim to:
+- support arbitrary dependencies graphs between containers
+- act as a security control to enforce that pod containers only run while the
+  sidecar is healthy. Restart of sidecar containers is a best effort
+- (alpha) support containers termination ordering
```

Comment:
A relatively common problem I heard over the years was the following:

> More effectively separate access to privileged / secure resources between containers. Filesystem resources, secrets, or more privileged process behaviors can be separated into a container that has higher access and is available before the regular serving/job container and until it has completed.

This is somewhat part of the service mesh use case (especially those that perform privileged iptables manipulation and hide access to secrets), but I don't see it called out here in some form as a goal (or a really good reason it's a non goal).  I want to make sure our design does not limit or constrain such use cases.  Other examples over the years that are not service mesh:

1. Run an agent that exchanges pod identity information to gain a time or space limited security credential (such as an SSL cert, remote token, whatever) alongside a web serving container that has no access to the pod identity credential, and thus can only expose the bound credential if compromised.
2. Start a FUSE daemon in a sidecar that mounts an object storage bucket into one of the shared pod emptydir volumes - have another container running with much lower privileges that can use that
3. Run a container image build with privileges in a sidecar container that responds to commands on a unix socket shared in a pod emptydir and takes commands from an unprivileged container to snapshot the state of the container.

If this was covered in an earlier discussion, would like to see details and at least a summary here.

Reply:
this was covered in discussions. What we get to in these discussions, that if there will be security boundaries support like this, it will equally apply to regular Init containers and introducing sidecars does not make this (already hard) problem harder to solve.

So we said we will postpone solving it applying it to all "infrastructure" containers. I'll add a note about it.

Reply:
I'm fine with a statement deferring it inline "we expect use cases to exist around security and we do not constrain them here (example of why we aren't constraining them)".  Just looking for enough to believe we aren't gonna be limited later based on decisions now.